File: temp_cleanup/lib/infrastructure/ai/agent/tools/tool_registry.dart
import '../models/agent_models.dart';

/// Abstract interface for all tools
/// Implementations must be stateless and thread-safe
abstract class AgentTool {
  /// Tool metadata (name, description, capabilities)
  ToolMetadata get metadata;

  /// Execute the tool with given parameters
  /// Returns result that should be JSON-serializable
  Future<dynamic> execute(Map<String, dynamic> parameters);

  /// Check if this tool can handle the given action
  bool canHandle(String action);

  /// Get required permissions for this tool
  List<String> getRequiredPermissions();

  /// Validate parameters before execution
  /// Throw exception if validation fails
  void validateParameters(Map<String, dynamic> parameters);
}

/// Registry for available tools
/// Manages tool discovery, validation, and execution
class ToolRegistry {
  final Map<String, AgentTool> _tools = {};
  final Map<String, List<String>> _capabilityIndex = {};
  final Map<String, List<String>> _domainIndex = {};

  /// Register a new tool
  void register(AgentTool tool) {
    _tools[tool.metadata.name] = tool;

    // Index capabilities
    for (final capability in tool.metadata.capabilities) {
      _capabilityIndex
          .putIfAbsent(capability, () => [])
          .add(tool.metadata.name);
    }

    // Index by domain (derived from tool name pattern)
    final domain = _extractDomain(tool.metadata.name);
    _domainIndex.putIfAbsent(domain, () => []).add(tool.metadata.name);
  }

  /// Unregister a tool
  bool unregister(String toolName) {
    final tool = _tools.remove(toolName);
    if (tool == null) return false;

    // Clean up indices
    for (final capability in tool.metadata.capabilities) {
      _capabilityIndex[capability]?.remove(toolName);
    }

    final domain = _extractDomain(toolName);
    _domainIndex[domain]?.remove(toolName);

    return true;
  }

  /// Get tool by name
  AgentTool? getTool(String toolName) => _tools[toolName];

  /// Find tools with specific capability
  List<AgentTool> findByCapability(String capability) {
    return (_capabilityIndex[capability] ?? [])
        .map((name) => _tools[name]!)
        .toList();
  }

  /// Find tools in a specific domain
  List<AgentTool> findByDomain(String domain) {
    return (_domainIndex[domain] ?? []).map((name) => _tools[name]!).toList();
  }

  /// Get all available tools
  List<AgentTool> getAllTools() => _tools.values.toList();

  /// Get all tool metadata (useful for agent planning)
  List<ToolMetadata> getAllMetadata() =>
      _tools.values.map((t) => t.metadata).toList();

  /// Find tools that can handle the given action
  List<AgentTool> findByAction(String action) {
    return _tools.values.where((tool) => tool.canHandle(action)).toList();
  }

  /// Execute a tool by name
  /// Validates parameters before execution
  Future<dynamic> executeTool(
    String toolName,
    Map<String, dynamic> parameters,
  ) async {
    final tool = _tools[toolName];
    if (tool == null) {
      throw ToolNotFoundException('Tool not found: $toolName');
    }

    try {
      tool.validateParameters(parameters);
      return await tool.execute(parameters);
    } catch (e) {
      throw ToolExecutionException(
        'Failed to execute tool $toolName: $e',
        toolName,
        e,
      );
    }
  }

  /// Get capabilities provided by all registered tools
  Set<String> getAllCapabilities() => _capabilityIndex.keys.toSet();

  /// Get all domains with registered tools
  Set<String> getAllDomains() => _domainIndex.keys.toSet();

  /// Check if tool is registered
  bool hasToolWithName(String toolName) => _tools.containsKey(toolName);

  /// Check if capability is available
  bool hasCapability(String capability) =>
      _capabilityIndex.containsKey(capability) &&
      _capabilityIndex[capability]!.isNotEmpty;

  /// Check if all required tools are available
  bool hasAllTools(List<String> toolNames) =>
      toolNames.every((name) => _tools.containsKey(name));

  /// Check if all required capabilities are available
  bool hasAllCapabilities(List<String> capabilities) =>
      capabilities.every((cap) => hasCapability(cap));

  /// Get number of registered tools
  int get toolCount => _tools.length;

  /// Clear all registered tools (useful for testing)
  void clear() {
    _tools.clear();
    _capabilityIndex.clear();
    _domainIndex.clear();
  }

  /// Extract domain from tool name (e.g., "ui_validation" -> "ui")
  static String _extractDomain(String toolName) {
    final parts = toolName.split('_');
    return parts.isNotEmpty ? parts.first : 'general';
  }
}

/// Exception thrown when tool is not found
class ToolNotFoundException implements Exception {
  final String message;

  ToolNotFoundException(this.message);

  @override
  String toString() => message;
}

/// Exception thrown during tool execution
class ToolExecutionException implements Exception {
  final String message;
  final String toolName;
  final Object? originalError;

  ToolExecutionException(this.message, this.toolName, this.originalError);

  @override
  String toString() => message;
}

/// Extension for easier tool lookup by capability
extension ToolRegistryCapabilityExt on ToolRegistry {
  /// Fluent API for finding tools by multiple capabilities
  List<AgentTool> findByCapabilities(List<String> capabilities) {
    if (capabilities.isEmpty) return [];

    final toolsByCapability =
        capabilities.map((cap) => findByCapability(cap).toSet()).toList();

    // Return tools that have at least one of the capabilities
    return toolsByCapability
        .reduce((acc, current) => acc.union(current))
        .toList();
  }
}

File: micro/backend/infrastructure/tools/tool_registry.py
"""
Tool Registry - Dynamic tool discovery and management.
Based on ADR-002: Zero hardcoded agent classes.
"""
import logging
from typing import Dict, List, Optional, Set
from domain.entities import ToolMetadata


logger = logging.getLogger(__name__)


class Tool:
    """Base class for all tools"""
    
    @property
    def metadata(self) -> ToolMetadata:
        """Return tool metadata"""
        raise NotImplementedError
    
    async def execute(self, parameters: Dict) -> any:
        """Execute the tool with given parameters"""
        raise NotImplementedError
    
    def can_handle(self, action: str) -> bool:
        """Check if this tool can handle the given action"""
        raise NotImplementedError
    
    def get_required_permissions(self) -> List[str]:
        """Get required permissions for this tool"""
        return self.metadata.required_permissions
    
    def validate_parameters(self, parameters: Dict) -> bool:
        """Validate parameters before execution"""
        return True


class ToolRegistry:
    """
    Registry for dynamic tool discovery and selection.
    
    Key Features:
    - Capability-based tool indexing
    - Domain-based organization
    - Zero hardcoded agent classes
    """
    
    def __init__(self):
        self._tools: Dict[str, Tool] = {}
        self._capability_index: Dict[str, List[str]] = {}
        self._domain_index: Dict[str, List[str]] = {}
    
    def register(self, tool: Tool) -> None:
        """Register a new tool"""
        metadata = tool.metadata
        self._tools[metadata.name] = tool
        
        # Index by capabilities
        for capability in metadata.capabilities:
            if capability not in self._capability_index:
                self._capability_index[capability] = []
            self._capability_index[capability].append(metadata.name)
        
        # Index by domain
        domain = metadata.domain
        if domain not in self._domain_index:
            self._domain_index[domain] = []
        self._domain_index[domain].append(metadata.name)
        
        logger.info(f"Registered tool: {metadata.name} with capabilities {metadata.capabilities}")
    
    def unregister(self, tool_name: str) -> bool:
        """Unregister a tool"""
        if tool_name not in self._tools:
            return False
        
        tool = self._tools.pop(tool_name)
        metadata = tool.metadata
        
        # Remove from capability index
        for capability in metadata.capabilities:
            if capability in self._capability_index:
                self._capability_index[capability].remove(tool_name)
        
        # Remove from domain index
        if metadata.domain in self._domain_index:
            self._domain_index[metadata.domain].remove(tool_name)
        
        logger.info(f"Unregistered tool: {tool_name}")
        return True
    
    def get_tool(self, tool_name: str) -> Optional[Tool]:
        """Get a tool by name"""
        return self._tools.get(tool_name)
    
    def get_tools_for_capabilities(self, capabilities: List[str]) -> List[Tool]:
        """Get all tools that support given capabilities"""
        tool_names: Set[str] = set()
        
        for capability in capabilities:
            if capability in self._capability_index:
                tool_names.update(self._capability_index[capability])
        
        return [self._tools[name] for name in tool_names if name in self._tools]
    
    def get_tools_for_domain(self, domain: str) -> List[Tool]:
        """Get all tools in a domain"""
        if domain not in self._domain_index:
            return []
        
        return [self._tools[name] for name in self._domain_index[domain]]
    
    def get_all_capabilities(self) -> List[str]:
        """Get all available capabilities"""
        return list(self._capability_index.keys())
    
    def get_all_domains(self) -> List[str]:
        """Get all available domains"""
        return list(self._domain_index.keys())
    
    def get_all_tools(self) -> List[Tool]:
        """Get all registered tools"""
        return list(self._tools.values())
    
    def get_tool_metadata(self, tool_name: str) -> Optional[ToolMetadata]:
        """Get metadata for a tool"""
        tool = self.get_tool(tool_name)
        return tool.metadata if tool else None
    
    def list_tools_by_capability(self, capability: str) -> List[str]:
        """List tool names that support a capability"""
        return self._capability_index.get(capability, [])
